Buoyancy updates

quarkytale opened the pull request:
Time for an update with the new values of WEC properties, see #33 for comparison.

- [x] Make buoyancy uniform
- [x] Disable for buoy link, enable for specific links - need for heave cone + PTO only
- [x] Dummy link for PTO's buoyancy
- [x] Set correct density
- [ ] Recalibrate volume, CoM, CoB vs CoG

quarkytale commented:
Made the buoyancy uniform and disabled for all links except heave cone and PTO. Behavior change now as the Buoy link doesn't have buoyancy the entire body kinda sinks slowly (expected?).

Need some values to move forward:
- Final volumes of the links
- What should I set the water density to be?

mabelzhang commented:
From meeting Andy: density = 1025

andermi commented on buoy_gazebo/gazebo/server.config:
```diff
@@ -16,18 +16,14 @@
-      </graded_buoyancy>
-    </plugin>
+      <uniform_fluid_density>1025</uniform_fluid_density>
+      <enable>MBARI_WEC_ROS::PTO_Buoyancy</enable>
+      <enable>MBARI_WEC_ROS::HeaveCone</enable>
```
```suggestion
      <enable>MBARI_WEC_BASE::PTO_Buoyancy</enable>
      <enable>MBARI_WEC_BASE::HeaveCone</enable>
```

andermi replied:
@quarkytale @hamilton8415 Try this...

quarkytale replied:
Something I was trying to verify if the plugin was acting upon the desired links.

andermi commented on buoy_gazebo/worlds/mbari_wec.sdf:
```diff
@@ -7,6 +7,21 @@
       <real_time_factor>1.0</real_time_factor>
     </physics>
 
+    <plugin
+      filename="gz-sim-buoyancy-system"
+      name="gz::sim::systems::Buoyancy">
+      <graded_buoyancy>
+        <default_density>1025</default_density>
+        <density_change>
+          <above_depth>0</above_depth>
+          <density>1</density>
+        </density_change>
+      </graded_buoyancy>
+      <!-- <uniform_fluid_density>1025</uniform_fluid_density>
+      <enable>MBARI_WEC_ROS::PTO_Buoyancy</enable>
+      <enable>MBARI_WEC_ROS::HeaveCone</enable> -->
+    </plugin>
```
@quarkytale @hamilton8415 how did this get here?
```suggestion
```

hamilton8415 replied:
Hail mary attempt from Dharini I think.  I too was surprised to see the graded_buoyancy tag make a re-appearance.

quarkytale replied:
Testing with buoyancy out of the `server.config` into the world files, hence the commented out plugin tags

andermi commented:
@quarkytale @hamilton8415 Sorry, I'm a bit confused by the diffs, but I believe you should be using

```
      <enable>MBARI_WEC_BASE::PTO_Buoyancy</enable>
      <enable>MBARI_WEC_BASE::HeaveCone</enable>
```

for the enables... `MBARI_WEC_BASE` not `MBARI_WEC_ROS`

quarkytale commented:
I updated the dummy link geometry close to the real values and was trying a few things out (commented on the respective difs). Updated those here as the current state of work and for further discussion.

hamilton8415 commented:
No, I had the dummy link at the correct values, nominally low mass and a geometry that matched the displaced volume of the PTO.  The most recent commit also changed of the mass of the  dummy buoyancy link to 605kg, this makes little sense for something we want to float.  

I think we should probably have a discussion in person to figure out what's going on and how to proceed and get this important feature working.  Arbitrarily changing values isn't going to get us very far.

quarkytale commented:
Sorry about that, reverted the values and cleaned up.

Still debugging why the link doesn't float, currently trying to trace from this debug statement where the Buoyancy seems to be loaded for just one entity
`[Dbg] [SystemManager.cc:74] Loaded system [gz::sim::systems::Buoyancy] for entity [1]`
where [1] corresponds to `Buoy` and [7] `HeaveCone`

quarkytale commented:
Is this what you're expecting to see @hamilton8415 ?

https://user-images.githubusercontent.com/24695820/223947115-4a1b2773-4c65-4010-a9e6-b1b98e501dee.mp4

quarkytale commented:
Does [BouyancyEngine](https://gazebosim.org/api/sim/7/classgz_1_1sim_1_1systems_1_1BuoyancyEnginePlugin.html) seem relevant? With min/max_volume tags it might remove the need for a dummy link?

hamilton8415 commented:
Listing the entity_name specifically has caused a buoyancy force to be applied for me, with MBARI_WEC_ROS.   Thank you for identifying that, I believe we can proceed and I will set the values appropriately from here.  After I commit that, please do not change values of the mass, collision shape/size, inertias, they are set to match the specific submerged volume and center of volume specified on the documentation site.  

I do not believe the buoyancy_engine plugin is relevant here.   

I did a little testing and found that the collision mesh .stl being used in the PTO has a displaced volume of about 0.161m^3. (found by adjusting the link mass until it neither sinks or floats).  The buoyancy plugin does include the contribution of more than one collision volume, so if one knew the center of  displacement for the collision mesh .stl, one cold make up the rest of the needed buoyancy (it should be .210m^3 total) and eliminate the need for the dummy link.  I may look into this.

andermi commented:
> if one knew the center of displacement for the collision mesh

The center of volume of the AABB for the pto_collision.stl mesh is -7.7723m in Z from the CAD/Link origin

andermi commented:
> should be .210m^3 total

did you mean `0.205m^3` ?

hamilton8415 commented:
Yes, .205, sorry about that mistake.  Looks correct that the remaining should be 0.044, but I'd think it needs to be centered quite a bit higher to offset the -7.7m diplacement of the 0.165m^3 component.  It's going to be a weighted average thing to land the resultant buoyancy force at -3m...  Also, the 0.161m^3 number can take a little more refinement.

andermi commented:
you're right, I had a math error... fixed above

hamilton8415 commented:
I'm happy to say I believe we now have the buoyancy forces correctly acting on the PTO and heave cone.  Run and have a look if you like, I realized a bit late in the game that the gui window reports the volume and center of volume being used, so that was helpful.  @andermi and I spoke and we think we will merge this straight into the fs_hydrodynamics branch, we can then sort the tests and get it all into main.
